File: backend/src/models/Policy.js
const mongoose = require('mongoose');
const mongoosePaginate = require('mongoose-paginate-v2');
const baseSchemaFields = require('./BaseSchema');

const policySchema = new mongoose.Schema({
  ...baseSchemaFields,
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true,
  },
  policyNumber: { type: String, trim: true, maxlength: 255 },
  policyName: { type: String, trim: true, maxlength: 255 },
  insuranceCompany: { type: String, trim: true, maxlength: 255 },
  policyType: { type: String, trim: true, maxlength: 255 },
  policyStartDate: { type: Date },
  policyEndDate: { type: Date },
  coverageAmount: { type: Number, min: 0 },
  status: { type: String, default: 'Active', trim: true, maxlength: 50 },
  gridFsFileId: { type: mongoose.Schema.Types.ObjectId, ref: 'fs.files' },
  originalFileName: { type: String, trim: true, maxlength: 255 },
  mimeType: { type: String, trim: true, maxlength: 255 },
  fileSize: { type: Number, min: 0 },
  sequenceNumber: { type: Number, min: 0 },
  agreement: {
    accepted: { type: Boolean },
    termsAccepted: { type: Boolean },
    termsVersion: { type: String, trim: true, maxlength: 50 },
    acceptedAt: { type: Date },
    appVersion: { type: String, trim: true, maxlength: 50 },
    platform: { type: String, trim: true, maxlength: 50 },
  },
}, { timestamps: true, strict: true });

// Compound index for optimal dashboard counts and pagination
policySchema.index({ userId: 1, isDeleted: 1, createdAt: -1 });

policySchema.pre('save', async function(next) {
  if (this.isNew) {
    const Counter = require('./Counter');
    const counter = await Counter.findOneAndUpdate(
      { userId: this.userId, entityType: 'Policy' },
      { $inc: { seq: 1 } },
      { new: true, upsert: true }
    );
    this.sequenceNumber = counter.seq;
  }
  next();
});

policySchema.plugin(mongoosePaginate);

module.exports = mongoose.model('Policy', policySchema);

File: mobile-app/lib/features/policies/data/models/policy.dart
class Policy {
  final String id;
  final String policyNumber;
  final String policyName;
  final String insuranceCompany;
  final String? policyType;
  final DateTime? policyStartDate;
  final DateTime? policyEndDate;
  final double? coverageAmount;
  final String status;
  final String? gridFsFileId;
  final String? originalFileName;
  final String? mimeType;
  final int? fileSize;
  final int? sequenceNumber;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  String get displayId {
    if (sequenceNumber != null) {
      return 'PCY${sequenceNumber.toString().padLeft(4, '0')}';
    }
    return id.substring(0, 8).toUpperCase();
  }

  Policy({
    required this.id,
    required this.policyNumber,
    required this.policyName,
    required this.insuranceCompany,
    this.policyType,
    this.policyStartDate,
    this.policyEndDate,
    this.coverageAmount,
    required this.status,
    this.gridFsFileId,
    this.originalFileName,
    this.mimeType,
    this.fileSize,
    this.sequenceNumber,
    this.createdAt,
    this.updatedAt,
  });

  factory Policy.fromJson(Map<String, dynamic> json) {
    return Policy(
      id: json['_id']?.toString() ?? '',
      policyNumber: json['policyNumber'] ?? '',
      policyName: json['policyName'] ?? '',
      insuranceCompany: json['insuranceCompany'] ?? '',
      policyType: json['policyType'],
      policyStartDate: json['policyStartDate'] != null ? DateTime.parse(json['policyStartDate']) : null,
      policyEndDate: json['policyEndDate'] != null ? DateTime.parse(json['policyEndDate']) : null,
      coverageAmount: json['coverageAmount']?.toDouble(),
      status: json['status'] ?? 'Active',
      gridFsFileId: json['gridFsFileId'],
      originalFileName: json['originalFileName'],
      mimeType: json['mimeType'],
      fileSize: (json['fileSize'] as num?)?.toInt(),
      sequenceNumber: (json['sequenceNumber'] as num?)?.toInt(),
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'policyNumber': policyNumber,
      'policyName': policyName,
      'insuranceCompany': insuranceCompany,
      'policyType': policyType,
      'policyStartDate': policyStartDate?.toIso8601String(),
      'policyEndDate': policyEndDate?.toIso8601String(),
      'coverageAmount': coverageAmount,
      'status': status,
      'gridFsFileId': gridFsFileId,
      'originalFileName': originalFileName,
      'mimeType': mimeType,
      'fileSize': fileSize,
    };
  }
}
